scaletest/trafficgen/run.go

```diff
@@ -0,0 +1,281 @@
+	)
+
+	logger.Info(ctx, "config",
+		slog.F("agent_id", agentID),
+		slog.F("reconnect", reconnect),
+		slog.F("height", height),
+		slog.F("width", width),
+		slog.F("tick_interval", tickInterval),
+		slog.F("bytes_per_tick", bytesPerTick),
+	)
+
+	// Set a deadline for stopping the text.
+	start := time.Now()
+	deadlineCtx, cancel := context.WithDeadline(ctx, start.Add(r.cfg.Duration))
+	defer cancel()
+	logger.Debug(ctx, "connect to workspace agent", slog.F("agent_id", agentID))
+
+	conn, err := r.client.WorkspaceAgentReconnectingPTY(ctx, codersdk.WorkspaceAgentReconnectingPTYOpts{
+		AgentID:   agentID,
+		Reconnect: reconnect,
+		Height:    height,
+		Width:     width,
+		Command:   "/bin/sh",
+	})
+	if err != nil {
+		logger.Error(ctx, "connect to workspace agent", slog.F("agent_id", agentID), slog.Error(err))
+		return xerrors.Errorf("connect to workspace: %w", err)
+	}
+
+	go func() {
+		<-deadlineCtx.Done()
+		logger.Debug(ctx, "close agent connection", slog.F("agent_id", agentID))
+		_ = conn.Close()
+	}()
+
+	// Wrap the conn in a countReadWriter so we can monitor bytes sent/rcvd.
+	crw := countReadWriter{ReadWriter: conn, ctx: deadlineCtx}
+
+	// Create a ticker for sending data to the PTY.
+	tick := time.NewTicker(tickInterval)
+	defer tick.Stop()
+
+	// Now we begin writing random data to the pty.
+	rch := make(chan error, 1)
+	wch := make(chan error, 1)
+
+	go func() {
+		<-deadlineCtx.Done()
+		logger.Debug(ctx, "context deadline reached", slog.F("duration", time.Since(start)))
+	}()
+
+	// Read forever in the background.
+	go func() {
+		logger.Debug(ctx, "reading from agent", slog.F("agent_id", agentID))
+		rch <- drainContext(deadlineCtx, &crw)
+		logger.Debug(ctx, "done reading from agent", slog.F("agent_id", agentID))
+		conn.Close()
+		close(rch)
+	}()
+
+	// Write random data to the PTY every tick.
+	go func() {
+		logger.Debug(ctx, "writing to agent", slog.F("agent_id", agentID))
+		wch <- writeRandomData(deadlineCtx, &crw, bytesPerTick, tick.C)
+		logger.Debug(ctx, "done writing to agent", slog.F("agent_id", agentID))
+		close(wch)
+	}()
+
+	// Write until the context is canceled.
+	if wErr := <-wch; wErr != nil {
+		return xerrors.Errorf("write to pty: %w", wErr)
+	}
+	if rErr := <-rch; rErr != nil {
+		return xerrors.Errorf("read from pty: %w", rErr)
+	}
+
+	duration := time.Since(start)
+
+	logger.Info(ctx, "results",
+		slog.F("duration", duration),
+		slog.F("sent", crw.BytesWritten()),
+		slog.F("rcvd", crw.BytesRead()),
+	)
+
+	return nil
+}
+
+// Cleanup does nothing, successfully.
+func (*Runner) Cleanup(context.Context, string) error {
+	return nil
+}
+
+// drainContext drains from src until it returns io.EOF or ctx times out.
+func drainContext(ctx context.Context, src io.Reader) error {
+	errCh := make(chan error, 1)
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				_, err := io.CopyN(io.Discard, src, 1)
```

Comment:
reading one byte at a time is likely to artificially limit the speed at which we can read, which in turn, will put back pressure over the network and to the shell, and limit write speed.

The reason that reading one byte at a time is slower is that reading from a network socket involves making a syscall, interrupting the userspace thread and switching over to kernel code execution.  There is a non-trivial amount of overhead in this operation, so it's best for performance to read with largish buffers.

I _suspect_ that here you're trying to ensure that this function will actually return when the context expires.  Firstly, I think that's unnecessary.  As mentioned above, we close the net.Conn so we should hit EOF or an ErrClosed pretty quickly.  Secondly, it could still block: if you call Read() with any non-zero buffer, including a single byte, that call could block if there's no data ready to be read.

Reply:
Now it's just a straight-up `io.Copy`.